services/processor/native/adapter/native_compiler_test.go

```diff
@@ -32,18 +32,30 @@ func TestCompileValidContract(t *testing.T) {
 		if testing.Short() {
 			t.Skip("Skipping compilation of contracts in short mode")
 		}
-		// give the test one minute timeout to compile
-		ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
-		defer cancel()
+		err := compilationOfMockCounterContract(t, parent)
+		require.NoError(t, err)
+	})
+}
 
-		cfg, cleanup := adapterTest.NewConfigWithTempDir(t)
-		defer cleanup()
-		compiler := NewNativeCompiler(cfg, parent.Logger, metric.NewRegistry())
+func TestCompileContractConcurrently(t *testing.T) {
+	with.Logging(t, func(parent *with.LoggingHarness) {
+		if testing.Short() {
+			t.Skip("Skipping compilation of contracts in short mode")
+		}
 
-		code := string(contracts.NativeSourceCodeForCounter(contracts.MOCK_COUNTER_CONTRACT_START_FROM))
-		_, err := compiler.Compile(ctx, code)
-		t.Log("Error : ", err)
-		require.NoError(t, err, "compile should not fail")
+		const concurrentCount = 5
+		doneChan := make(chan interface{}, concurrentCount)
+		for i := 0; i < concurrentCount; i++ {
+			go func() {
+				err := compilationOfMockCounterContract(t, parent)
```

Comment:
You can `require.NoError(err)` here and just pass nil in the channel, then read all the nils in line 55, a bit more to my taste :)

Reply:
`require.NoError` calls `runtime.Goexit()` when an error is detected. this means we have to be extra careful in how we signal into `doneChan` and `defer` the signaling so it is executed in case an error is detected.

I accept the comment, but it is a little fragile. This is why I normally prefer to `require` only from the top goroutine